src/audio/smart_amp_test.c

```diff
@@ -0,0 +1,722 @@
+		 0x89, 0xf1, 0x00, 0x0c, 0x29, 0xce, 0x16, 0x35);
+
+DECLARE_TR_CTX(smart_amp_comp_tr, SOF_UUID(smart_amp_comp_uuid), LOG_LEVEL_INFO);
+
+struct smart_amp_data {
+	struct sof_smart_amp_config config;
+	struct smart_amp_model_data model;
+
+	struct comp_buffer *source_buf; /**< stream source buffer */
+	struct comp_buffer *feedback_buf; /**< feedback source buffer */
+	struct comp_buffer *sink_buf; /**< sink buffer */
+
+	uint32_t in_channels;
+	uint32_t out_channels;
+};
+
+static void free_mem_load(struct smart_amp_data *sad)
+{
+	if (!sad) {
+		comp_cl_err(&comp_smart_amp, "free_mem_load(): invalid sad");
+		return;
+	}
+
+	if (sad->model.data) {
+		rfree(sad->model.data);
+		sad->model.data = NULL;
+		sad->model.data_size = 0;
+		sad->model.crc = 0;
+		sad->model.data_pos = 0;
+	}
+}
+
+static int alloc_mem_load(struct smart_amp_data *sad, uint32_t size)
+{
+	if (!size)
+		return 0;
+
+	if (!sad) {
+		comp_cl_err(&comp_smart_amp, "alloc_mem_load(): invalid sad");
+		return -EINVAL;
+	}
+
+	free_mem_load(sad);
+
+	sad->model.data = rballoc(0, SOF_MEM_CAPS_RAM, size);
+
+	if (!sad->model.data) {
+		comp_cl_err(&comp_smart_amp, "alloc_mem_load() alloc failed");
+		return -ENOMEM;
+	}
+
+	bzero(sad->model.data, size);
+	sad->model.data_size = size;
+	sad->model.data_pos = 0;
+
+	return 0;
+}
+
+static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
+				      struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_process *sa;
+	struct sof_ipc_comp_process *ipc_sa =
+		(struct sof_ipc_comp_process *)comp;
+	struct smart_amp_data *sad;
+	struct sof_smart_amp_config *cfg;
+	size_t bs;
+	int ret;
+
+	comp_cl_info(&comp_smart_amp, "smart_amp_new()");
+
+	dev = comp_alloc(drv, COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+	dev->drv = drv;
+
+	sa = COMP_GET_IPC(dev, sof_ipc_comp_process);
+
+	ret = memcpy_s(sa, sizeof(*sa), ipc_sa,
+		       sizeof(struct sof_ipc_comp_process));
+	assert(!ret);
+
+	sad = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*sad));
+
+	if (!sad) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, sad);
+
+	cfg = (struct sof_smart_amp_config *)ipc_sa->data;
+	bs = ipc_sa->size;
+
+	if ((bs > 0) && (bs < sizeof(struct sof_smart_amp_config))) {
+		comp_err(dev, "smart_amp_new(): failed to apply config");
+
+		if (sad)
+			rfree(sad);
+		rfree(sad);
+		return NULL;
+	}
+
+	memcpy_s(&sad->config, sizeof(struct sof_smart_amp_config), cfg, bs);
+
+	dev->state = COMP_STATE_READY;
+
+	return dev;
+}
+
+static int smart_amp_set_config(struct comp_dev *dev,
+				struct sof_ipc_ctrl_data *cdata)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct sof_smart_amp_config *cfg;
+	size_t bs;
+
+	comp_info(dev, "smart_amp_set_config()");
+
+	/* Copy new config, find size from header */
+	cfg = (struct sof_smart_amp_config *)cdata->data->data;
+	bs = cfg->size;
+
+	comp_info(dev, "smart_amp_set_config(), blob size = %u", bs);
```

Comment:
Logging specific params like blob size seems useful (however pls consider dbg level) while not sure about the previous one - seems duplicated?

Reply:
I've changed log level to dbg and added actual and expected blob size log.